### PR TITLE
Simplify templated math tests

### PR DESCRIPTION
**Pull request by BenjaminW3**

### CODE REVIEW

File: test/unit/math/src/math.cpp [hunk 1]
```diff
@@ -224,30 +225,9 @@ TEST_CASE("mathOps", "[math] [operator]")
...
-
-    alpaka::meta::forEachType< TestAccs >(
-        ForEachFunctor< float >( ),
-        seed
-    );
+    using Acc = std::tuple_element_t<0u, TestType>;
+    using Functor = std::tuple_element_t<1u, TestType>;
+
+    alpaka::meta::forEachType< DataTypes >(
+        TestTemplate< Acc, Functor >());
```
Comment (BenjaminW3) [inline test/unit/math/src/math.cpp:232]: It is possible to use:
```
alpaka::meta::CartesianProduct<std::tuple, TestAccs, DataTypes, Functors>
```
but then nvcc complains about: `error #456: excessive recursion at instantiation of class`.
Therefore we still need one indirection through `forEachType`.

File: test/unit/math/src/math.cpp [hunk 2]
```diff
@@ -42,19 +66,25 @@ struct TestKernel
...
 struct TestTemplate
 {
-    template < typename TFunctor >
-    auto operator() ( unsigned long seed ) -> void
+    template<
+        typename TData>
+    auto operator()() -> void
     {
+        std::random_device rd{};
+        auto const seed = rd();
```
Comment (BenjaminW3) [inline test/unit/math/src/math.cpp:80]: We now randomly initialize the seed to get more coverage over time. I copied those two lines from the axpy test.

File: test/unit/math/src/math.cpp [hunk 3]
```diff
@@ -15,8 +15,32 @@
...
+        alpaka::test::unit::math::UnaryFunctors,
+        alpaka::test::unit::math::BinaryFunctors
+    >;
+
+using TestAccFunctorTuples =
+    alpaka::meta::CartesianProduct<
+        std::tuple,
+        TestAccs,
+        Functors
+    >;
```
Comment (BenjaminW3) [inline test/unit/math/src/math.cpp:37]: @jkrude Here is the meta programming that generates an entry per functor per accelerator.
You can do something similar for your use case e.g.:
```
using TestAccDataTypeTuples =
    alpaka::meta::CartesianProduct<
        std::tuple,
        alpaka::test::acc::TestAccs,
        DataTypes
    >;
```
together with `TEMPLATE_LIST_TEST_CASE` and `std::tuple_element_t` you can iterate over and get the current types.

Comment (BenjaminW3) [reply test/unit/math/src/math.cpp:37]: Depending on the size of template lists you may get issues with the template instantiation depth in nvcc as I did here. So you may still need one layer of `forEachType`.

**Review by sbastrakov — APPROVED** [review]

**Pull request merged**